src/content/posts/2024/09/20/xcode-plane.mdx

```diff
@@ -0,0 +1,102 @@
+
+## Sharing code
+
+Sooner or later in the life of the project, you need to share code within your app, whether through frameworks or libraries you've created to use with extensions, or to enforce access boundaries and improve your app's architecture. You'll also likely rely on third-party packages from other developers. You can think of your project as a [directed acyclic graph (DAG)](https://en.wikipedia.org/wiki/Directed_acyclic_graph), where each node represents a module—often modeled by an Xcode target, like a library or framework. Now, here's the challenge: **managing this graph is hard**. Adding or removing a module can create ripple effects throughout the entire system. For example, a dynamic framework might require certain upstream targets to embed it, or transitive static symbols leaking through a dynamic module could force their Swift module interfaces to be exposed elsewhere in the graph.
+
+If this sounds too abstract, think of it this way: *you can't just drag and drop modules in and out of the project without consequences.* With one or two modules, you might get by, but as the graph grows, so does the complexity of the relationships between modules—and with it, the risk of breaking something. Some developers we've spoken to have described this issue as “needing to learn how Xcode works.” But we don't think developers should have to. **Modularization is a standard requirement in almost every project, and it should be simple by default.**
+
+Did you know that your project might compile successfully purely due to side effects from previous compilation steps? One of your targets might resolve an import from another target not because it's declared as a dependency, but because it was already present in derived data.
+
+Then there are the common questions developers face: *Should this module be static or dynamic? Should it be a framework or a library? Can this product type embed dynamic frameworks? And what about dynamic libraries? What are the implications of linking a static library transitively?* These are questions we ask ourselves frequently, and often have to answer by reverse-engineering Xcode-generated projects and validating outputs with App Store Connect. Surprisingly, this crucial knowledge—rarely discussed in depth—can have serious implications for your app, from runtime crashes to unnecessarily bloating its size.
+
+In this case, the convenience of implicitness manifests as implicitly linked dependencies and implicitly resolved linking types (static or dynamic) through [mergeable libraries](https://developer.apple.com/documentation/xcode/configuring-your-project-to-use-mergeable-libraries). These are additional variables that introduce build-time complexity and variability, which can not only impact build performance but also make certain features, like Swift Previews, work inconsistently.
```

Comment:
"In this case, the convenience of implicitness manifests as implicitly linked dependencies and implicitly resolved"

"...the convenience of implicitness manifests as implicitly linked ..." -> I think this is worth rephrasing. The convenience of implicitness manifests as longer build times or build errors due to implicitly linked dependencies, not as "implicitly linked dependencies"